File: List/Petukhov/src/lnu/edu/ua/LinkedList.java
package lnu.edu.ua;

public class LinkedList<T> implements List<T> {

    private Node<T> head;
    private Node<T> tail;
    private int size;

    public LinkedList() {
        head = null;
        tail = null;
        size = 0;
    }

    @Override
    public void add(T value) {
        Node<T> node = new Node<>(value);
        if (head == null) {
            head = node;
            tail = node;
        } else {
            tail.setNext(node);
            node.setPrev(tail);
            tail = node;
        }
        size++;
    }

    @Override
    public void add(T value, int index) {
        if (index < 0 || index > size) {
            throw new IndexOutOfBoundsException();
        }
        Node<T> node = new Node<>(value);
        if (index == 0) {
            node.setNext(head);
            if (head != null) {
                head.setPrev(node);
            }
            head = node;
            if (tail == null) {
                tail = node;
            }
        } else if (index == size) {
            tail.setNext(node);
            node.setPrev(tail);
            tail = node;
        } else {
            Node<T> current = head;
            for (int i = 0; i < index; i++) {
                current = current.getNext();
            }
            node.setPrev(current.getPrev());
            node.setNext(current);
            current.getPrev().setNext(node);
            current.setPrev(node);
        }
        size++;
    }

    @Override
    public void addAll(List<T> list) {
        for (int i = 0; i < list.size(); i++) {
            add(list.get(i));
        }
    }

    @Override
    public T remove(int index) {
        if (index < 0 || index >= size) {
            throw new IndexOutOfBoundsException();
        }
        Node<T> current = head;
        for (int i = 0; i < index; i++) {
            current = current.getNext();
        }
        if (current.getPrev() != null) {
            current.getPrev().setNext(current.getNext());
        } else {
            head = current.getNext();
        }
        if (current.getNext() != null) {
            current.getNext().setPrev(current.getPrev());
        } else {
            tail = current.getPrev();
        }
        size--;
        return current.getValue();
    }

    @Override
    public T remove(T element) {
        Node<T> current = head;
        while (current != null) {
            if (current.getValue().equals(element)) {
                if (current.getPrev() != null) {
                    current.getPrev().setNext(current.getNext());
                } else {
                    head = current.getNext();
                }
                if (current.getNext() != null) {
                    current.getNext().setPrev(current.getPrev());
                } else {
                    tail = current.getPrev();
                }
                size--;
                return current.getValue();
            }
            current = current.getNext();
        }
        return null;
    }

    @Override
    public T get(int index) {
        if (index < 0 || index >= size) {
            throw new IndexOutOfBoundsException();
        }
        Node<T> current = head;
        for (int i = 0; i < index; i++) {
            current = current.getNext();
        }
        return current.getValue();
    }

    @Override
    public void set(T value, int index) {
        if (index < 0 || index >= size) {
            throw new IndexOutOfBoundsException();
        }
        Node<T> current = head;
        for (int i = 0; i < index; i++) {
            current = current.getNext();
        }
        current.setValue(value);
    }

    public int size() {
        return size;
    }

    @Override
    public boolean isEmpty() {
        return size == 0;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder("[");
        Node<T> current = head;
        while (current != null) {
            sb.append(current.getValue());
            if (current.getNext() != null) {
                sb.append(", ");
            }
            current = current.getNext();
        }
        sb.append("]");
        return sb.toString();
    }
}

File: List/LinkedListVasylyk/LinkedList/LinkedList.cs
namespace LinkedList;

public class CustomLinkedList<T>
{
    private Node<T> first;

    public CustomLinkedList(Node<T> first)
    {
        this.first = first;
    }

    public void AddAtEnd(T value)
    {
        Node<T> current = first;
        while (current.HasNext())
        {
            current = current.NextNode;
        }
        current.NextNode = new Node<T>(value, current, null);
    }

    public void AddAtBegin(T value)
    {
        Node<T> newNode = new Node<T>(value, null, first);
        first.PrevNode = newNode;
        first = newNode;
    }

    public void AddByIndex(int index, T value)
    {
        int i = 0;
        Node<T> current = first;
        while (current.HasNext() && i < index)
        {
            i++;
            current = current.NextNode;
        }
        if (i == index)
        {
            Node<T> newNode = new Node<T>(value, current.PrevNode, current);
            current.PrevNode.NextNode = newNode;
            current.PrevNode = newNode;
        }
    }
    public void Remove(T value)
    {
        Node<T> current = first;
        while (current != null)
        {
            if (current.Value.Equals(value))
            {
                if (current.PrevNode != null)
                {
                    current.PrevNode.NextNode = current.NextNode;
                }
                else
                {
                    first = current.NextNode;
                }

                if (current.NextNode != null)
                {
                    current.NextNode.PrevNode = current.PrevNode;
                }
                return;
            }
            current = current.NextNode;
        }
    }

    public override string ToString()
    {
        Node<T> current = first;
        string output = current.Value + " ";
        while (current.HasNext())
        {
            current = current.NextNode;
            output += current.Value + " ";
        }
        return output;
    }
}